src/memory/AmdFlash.hh

```diff
@@ -107,17 +124,39 @@ public:
 			uint8_t bootBlockFlag = 0;
 			uint8_t programSuspend = 0;
 		} primaryAlgorithm = {};
+
+		constexpr void validate() const {
+			assert(!command || primaryAlgorithm.version.major == 1);
+		}
 	};
 	struct Misc {
 		bool statusCommand : 1 = false;
 		bool continuityCommand : 1 = false;
+
+		constexpr void validate() const {
+			assert(!continuityCommand || statusCommand);
+		}
 	};
 	struct Chip {
 		AutoSelect autoSelect;
 		Geometry geometry;
 		Program program = {};
 		CFI cfi = {};
 		Misc misc = {};
+
+		constexpr void validate() const {
+			autoSelect.validate();
+			geometry.validate();
+			program.validate();
+			cfi.validate();
+			misc.validate();
+		}
+	};
+	struct ValidatedChip {
+		constexpr ValidatedChip(Chip&& chip_) : chip(std::move(chip_)) {
+			chip.validate();
+		}
+		const Chip chip;
```

Comment:
Detail, background info, no need to change anything.
I understand why you want to store `chip` as a `const` member: after validation it may not be changed anymore.

On the other hand some c++ coding guidelines recommend to not use const member variables (*1). The reason for this is that it disables move-operations. E.g. see this experiment: https://godbolt.org/z/zoxfrPoef

In case of `{Validated,}Chip` this isn't relevant (moving and copying is the same). But I noticed anyway because you explicitly *move* `Chip` into `ValidatedChip`, (and then later *copy* it out into `AmdFlash`).


(*1) We don't always follow this guideline in openMSX. Partly because I only recently (in the last few years) learned about this. And partly because the tradeoff "allowing move" vs "more descriptive code" tilts towards the latter.

Reply:
Given the typical compile-time context it doesn’t really matter indeed. I’ll change the constructor to take `const Chip& chip_` though, so that it doesn’t give the false impression that it’s being moved.